### ISSUE TITLE
cli feature causes compilation error on latest

### ISSUE THREAD

**Issue by simon-wh** [tauri-apps/tauri#1188]
**Describe the bug**
When using the cli feature on latest tauri

**To Reproduce**
1. Write basic cli configuration to `tauri.conf.json` which enables the "cli" feature of tauri
2. Run `yarn tauri dev`
3. Watch as it errors out while compiling

```
error[E0308]: mismatched types
  --> /home/syte/.cargo/registry/src/github.com-1ecc6299db9ec823/tauri-api-0.7.5/src/cli.rs:77:7
   |
77 |   let (subcommand_name, subcommand_matches_option) = matches.subcommand();
   |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^   -------------------- this expression has type `std::option::Option<(&str, &ArgMatches)>`
   |       |
   |       expected enum `std::option::Option`, found tuple
   |
   = note: expected enum `std::option::Option<(&str, &ArgMatches)>`
             found tuple `(_, _)`

error[E0599]: no method named `required_unless` found for struct `Arg<'_>` in the current scope
   --> /home/syte/.cargo/registry/src/github.com-1ecc6299db9ec823/tauri-api-0.7.5/src/cli.rs:171:63
    |
171 |   clap_arg = bind_string_arg!(arg, clap_arg, required_unless, required_unless);
    |                                                               ^^^^^^^^^^^^^^^ help: there is an associated function with a similar name: `required_if_eq`

error[E0599]: no method named `required_unless` found for struct `Arg<'_>` in the current scope
   --> /home/syte/.cargo/registry/src/github.com-1ecc6299db9ec823/tauri-api-0.7.5/src/cli.rs:173:63
    |
173 |   clap_arg = bind_string_arg!(arg, clap_arg, required_unless, required_unless);
    |                                                               ^^^^^^^^^^^^^^^ help: there is an associated function with a similar name: `required_if_eq`

error[E0599]: no method named `required_unless_all` found for struct `Arg<'_>` in the current scope
   --> /home/syte/.cargo/registry/src/github.com-1ecc6299db9ec823/tauri-api-0.7.5/src/cli.rs:174:52
    |
174 |   clap_arg = bind_string_slice_arg!(arg, clap_arg, required_unless_all);
    |                                                    ^^^^^^^^^^^^^^^^^^^ method not found in `Arg<'_>`

error[E0599]: no method named `required_unless_one` found for struct `Arg<'_>` in the current scope
   --> /home/syte/.cargo/registry/src/github.com-1ecc6299db9ec823/tauri-api-0.7.5/src/cli.rs:175:52
    |
175 |   clap_arg = bind_string_slice_arg!(arg, clap_arg, required_unless_one);
    |                                                    ^^^^^^^^^^^^^^^^^^^ help: there is an associated function with a similar name: `required_unless_present`

error[E0599]: no method named `required_if` found for struct `Arg<'_>` in the current scope
   --> /home/syte/.cargo/registry/src/github.com-1ecc6299db9ec823/tauri-api-0.7.5/src/cli.rs:181:42
    |
181 |   clap_arg = bind_if_arg!(arg, clap_arg, required_if);
    |                                          ^^^^^^^^^^^ help: there is an associated function with a similar name: `requires_if`

error: aborting due to 6 previous errors

Some errors have detailed explanations: E0308, E0599.
For more information about an error, try `rustc --explain E0308`.
error: could not compile `tauri-api`
```

**Platform and Versions (please complete the following information):**

```
Operating System - Linux(5.10.12-arch1-1) - linux/x64

Node.js environment
  Node.js - 10.23.2
  tauri.js - 0.14.1

Rust environment
  rustc - 1.51.0-nightly
  cargo - 1.51.0-nightly
  tauri-bundler - 0.9.4

Global packages
  NPM - 6.14.11
  yarn - 1.22.10

App directory structure
/.git
/.github
/node_modules
/public
/src
/src-tauri
/wooting-analog-midi

App
  tauri.rs - 0.11.1
  mode - embedded-server
  build-type - bundle
  CSP - default-src blob: data: filesystem: ws: http: https: 'unsafe-eval' 'unsafe-inline'
  distDir - ../build
  devPath - http://localhost:3000
Done in 2.82s.
```

**Additional context**

**Stack Trace**

**Comment by dskkato** [comment]
I don't know why, but enabling `cli` feature triggers to use unintended `clap` version.

`tauri-api` its self is set to use `clap` with `version = "3.0.0-beta.1"` for cli feature:

https://github.com/tauri-apps/tauri/blob/77a29f7d8181adbd47b28234d3d7352e683ef452/tauri-api/Cargo.toml#L36

But, when cli is on, current latest release tauri version attempt to use `clap` with `3.0.0-beta.2` which is incompatible API.

https://github.com/dskkato/my-tauri-app/blob/45791729ae662ac688bfd179ae7ea6e4b0ba3c8b/src-tauri/Cargo.lock#L214-#216

Perhaps, minimum reproducible step after `tauri init` is to add cli config to `tauri.conf.json` as bellow.

https://github.com/dskkato/my-tauri-app/blob/45791729ae662ac688bfd179ae7ea6e4b0ba3c8b/src-tauri/tauri.conf.json#L7-L11

Note that I could not reproduce this compile error when I cloned this repo and built with `cargo +nightly build --features cli`.

**Comment by lucasfernog** [comment]
that's weird because our main test app also has the cli feature and it never had this error. Can you pin the version on your Cargo manifest? We're in the middle of a huge update 😃 it'll be the biggest tauri release until beta.

**Comment by dskkato** [comment]
I think this compile error occurs only when users (or cargo command) download tauri from crates.io. In that case, git dependencies in the tauri-api might be discarded according to Rust doc, 

https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-git-repositories

It says:

> Note: crates.io does not allow packages to be published with git dependencies (git dev-dependencies are ignored). See the Multiple locations section for a fallback alternative.

It also describes fallback alternative approach by specifying `version` along with `path` or `git`

https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#multiple-locations

But, the problems is that clap 3.0.0 (including any beta versions) has not been published yet, I'm not sure how Rust resolve clap dependencies. I mean, there is no fallback point since no 3.0.0-beta.1 is there at https://crates.io/crates/clap now. One possibility is that it uses latest clap-rs/clap where its head revision is in 3.0.0-beta.2.

~~Thankfully, one clap maintainer published clap-v3 as preview release as [clap-v3](https://crates.io/crates/clap-v3) at 3.0.0-beta.1. Using the clap-v3 might be more stable rather than depending on its repository directory.~~

I'm sorry, clap-v3 seems to have been published by non clap member, according to clap`s gitter log. So, using clap-v3 crate may also be unreliable.

**Comment by dskkato** [comment]
> But, the problems is that clap 3.0.0 (including any beta versions) has not been published yet, I'm not sure how Rust resolve clap dependencies. I mean, there is no fallback point since no 3.0.0-beta.1 is there at https://crates.io/crates/clap now. One possibility is that it uses latest clap-rs/clap where its head revision is in 3.0.0-beta.2.

Oops, there is [clap 3.0.0-beta.1](https://crates.io/crates/clap/3.0.0-beta.1)...

**Comment by dskkato** [comment]
[According to this post](https://users.rust-lang.org/t/specify-the-exact-version-in-cargo/24066), one possible way to avoid this problem may be to specify the clap version with something like `version = "=3.0.0-beta.1"` (with equal mark).

https://github.com/tauri-apps/tauri/blob/77a29f7d8181adbd47b28234d3d7352e683ef452/tauri-api/Cargo.toml#L36

**Comment by dskkato** [comment]
There seems to be two ways to resolve this compile error.

1. Add exact clap version info to Cargo.toml (user level)

    As is mentioned by @lucasfernog, pin clap version by writing to `src-tauri/Cargo.toml` as follows when using `cli` features. Equal mark is crucial here.

    ```toml
    [dependencies]
    clap = "=3.0.0-beta.1"
    ```

2. Fix `tauri/tauri-app` crate (library level)

    Specifying clap version as `3.0.0-beta.1` may introduce unintended crate update when clap is released `3.0.0-beta.2` or higher versions (current status), add equal mark here as  `=3.0.0-beta.1`, [according to this post](https://users.rust-lang.org/t/specify-the-exact-version-in-cargo/24066)

    https://github.com/tauri-apps/tauri/blob/77a29f7d8181adbd47b28234d3d7352e683ef452/tauri-api/Cargo.toml#L36

Please note that I cannot test 2nd method since it only happens when `tauri` was downloaded from crates.io. O think it requires crate update to test.

@lucasfernog, If the 2nd approach seems to be right way, shall I create a PR for that...? (though it will just add a equal mark)?

**Comment by lucasfernog** [comment]
Yeah you're right and we should have an equal mark in there, sorry about that :( I'd love the PR!
